doc/architectural/static-functions.md

```diff
@@ -0,0 +1,153 @@
+\ingroup module_hidden
+\page modular-verification-static Modular Verification of Static Functions
+
+\section verification-of-static Modular Verification of Static Functions
+
+\author Kareem Khazem
+
+This page describes how to use CBMC on static functions.
+
+\tableofcontents
+
+CBMC can check libraries and other codebases that expose several
+entry points. To do this, users typically write a *harness* that
+captures the entry points' API contract, and that calls into the API
+with unconstrained values. The example below shows such a library and
+harness:
+
+\code{.c}
+void public_api_function(const int *a, int *b)
+{
+  // ...
+  private_function(a, b);
+  // ...
+}
+
+static void private_function(const int *a, int *b)
+{
+  // ...
+}
+\endcode
+
+The harness sets up some assumptions and then calls into the API:
+
+\code{.c}
+void harness()
+{
+  int a, b;
+  __CPROVER_assume(a < 10);
+  __CPROVER_assume(a >= 0);
+  public_api_function(&a, &b);
+  __CPROVER_assert(b != a);
+}
+\endcode
+
+The following commands build and check this function:
+
+\code{.sh}
+> goto-cc -c -o library.o library.c
+> goto-cc -c -o harness.o harness.c
+> goto-cc -o to_check.gb library.o harness.o
+> cbmc to_check.gb
+\endcode
+
+\subsection stubbing-out-static Stubbing out static functions
+
+For performance reasons, it might be desirable to analyze the static
+function separately from the API function. We can analyze the API
+function by "stubbing out" the static function, replacing it with a
+function that does nothing apart from asserting that its inputs satisfy
+the function's contract. Add the following to `harness.c`:
+
+\code{.c}
+static void private_function(const int *a, int *b)
+{
+  __CPROVER_assert( /* ... */ );
+  __CPROVER_assert( /* ... */ );
+  // ...
+  __CPROVER_assume( /* ... */ );
+  __CPROVER_assume( /* ... */ );
+}
+\endcode
+
+And build as follows, stripping the original static function out of its
+object file:
+
+\code{.sh}
+> goto-cc -c -o library.o library.c
+> goto-instrument --remove-function-body private_function library.o library-no-private.o
+>
+> goto-cc -c -o harness.o harness.c
+>
+> goto-cc -o to_check.gb library-no-private.o harness.o
+> cbmc to_check.gb
+\endcode
+
+\subsection checking-static Separately checking static functions
+
+We should now also write a harness for `private_function`. However,
+since that function is marked `static`, it is not possible for functions
+in external files to call it. We can write and link a harness by
+stripping the `static` attribute from `private_function` using goto-cc's
+`--export-function-local-symbols` flag.
```

Comment:
Weird name-- a static function is a file-local symbol not function-local

Reply:
We chose this name after a long discussion on the [original PR](https://github.com/diffblue/cbmc/pull/4409)...I agree that file-local is more intuitive. I've deprecated the old flag (users now see a warning if they use it) and changed the docs and command-line interface to use `--export-file-local-symbols`. This is actually better, because the name-mangling had been using the prefix `__CPROVER_file_local` instead of `__CPROVER_function_local`, so it's all now more consistent.